src/s3path.jl

```diff
@@ -210,45 +228,73 @@ end
 function FilePathsBase.walkpath(fp::S3Path; kwargs...)
     # Select objects with that prefix
     objects = s3_list_objects(fp.config, fp.bucket, fp.key; delimiter="")
+    root = joinpath(fp, "/")
 
     # Construct a new Channel using a recursive internal `_walkpath!` function
-    return Channel(ctype=typeof(fp)) do chnl
-        _walkpath!(fp, fp, Iterators.Stateful(objects), chnl; kwargs...)
+    return Channel(ctype=typeof(fp), csize=128) do chnl
+        _walkpath!(root, root, Iterators.Stateful(objects), chnl; kwargs...)
     end
 end
 
 function _walkpath!(root::S3Path, prefix::S3Path, objects, chnl; topdown=true, onerror=throw, kwargs...)
+    @assert root.isdirectory
+    @assert prefix.isdirectory
+
     while true
         try
             # Start by inspecting the next element
-            next = Base.peek(objects)
+            obj = Base.peek(objects)
 
             # Early exit condition if we've exhausted the iterator or just the current prefix.
-            next === nothing && return nothing
-            startswith(next["Key"], prefix.key) || return nothing
+            obj === nothing && return nothing
 
             # Extract the non-root part of the key
-            k = chop(next["Key"], head=length(root.key), tail=0)
+            k = chop(obj["Key"], head=length(root.key), tail=0)
 
-            # Determine the next appropriate child path
-            # 1. Next is a direct descendant of the current prefix (ie: we have a prefix object)
-            # 2. Next is a distant descendant of the current prefix (ie: we don't have prefix objects)
             fp = joinpath(root, k)
             _parents = parents(fp)
-            child, recurse = if last(_parents) == prefix || fp.segments == prefix.segments
+
+            # If the filepath matches our prefix then pop it off and continue
+            # Cause we would have already processed it before recursing
+            child = if prefix.segments == fp.segments
                 popfirst!(objects)
-                fp, isdir(fp)
-            else
+                continue
+            # If the filpath is a direct descendant of our prefix then check if it
+            # is a directory too
+            elseif last(_parents) == prefix
+                popfirst!(objects)
+                # If our current path is a prefix for the next path then we can assume that
+                # the current path should be a directory without needing to call `isdir`
+                next = Base.peek(objects)
+                is_dir = (next !== nothing && startswith(next["Key"], fp.key)) || isdir(fp)
```

Comment:
We only call `isdir` when our current path isn't a prefix for the next. This optimization assumes `s3_list_objects` results will continue to be returned in topdown order.